RecoHGCal/TICL/plugins/HGCGraph.cc

```diff
@@ -24,17 +25,39 @@ void HGCGraph::makeAndConnectDoublets(const TICLLayerTiles &histo,
   isOuterClusterOfDoublets_.resize(layerClusters.size());
   allDoublets_.clear();
   theRootDoublets_.clear();
-  for (int zSide = 0; zSide < 2; ++zSide) {
+  for (const auto &r : regions) {
+    bool isGlobal = (r.index == -1);
+    auto zSide = r.zSide;
+    int startEtaBin, endEtaBin, startPhiBin, endPhiBin;
+
+    if (isGlobal) {
+      startEtaBin = 0;
+      startPhiBin = 0;
+      endEtaBin = nEtaBins;
+      endPhiBin = nPhiBins;
+    } else {
+      auto firstLayerOnZSide = maxNumberOfLayers * zSide;
+      const auto &firstLayerHisto = histo[firstLayerOnZSide];
+
+      int entryEtaBin = firstLayerHisto.etaBin(r.origin.eta());
+      int entryPhiBin = firstLayerHisto.phiBin(r.origin.phi());
+      startEtaBin = std::max(entryEtaBin - deltaIEta, 0);
+      endEtaBin = std::min(entryEtaBin + deltaIEta, nEtaBins);
+      startPhiBin = entryPhiBin - deltaIPhi;
+      endPhiBin = entryPhiBin + deltaIPhi;
+    }
```

Comment:
I'm not sure I follow the logic here. You open a window in [eta,phi] here around the impact point. But later on, at lines 65,66 and 70, we open it up again, once for each bin in the original [eta,phi] space. This will likely both kill the performance and will add too many clusters (unless the compatibility kills them all, in which case we only waste time).
I'd suggest a more conservative approach of having a single bin here (so +1 for both ends) and open the window later on.

Reply:
Hi @rovere, 
lines 57-59 define the outer loop. Previously it was by default over the full range 0-nEtaBins, 0-nPhiBins.
This remains the same for the global seeded region, and it is restricted to a window +/- 2 (as the following loop on the inner region, but could be changed) around the entry point on the first layer.
The window is needed to keep efficiency in case the first compatible 2dcl is few layers beyond the first, while the track is propagate only to the first HGCAL layer. +/-2 might be too much, but the only bin containing the entry position can also be too tight. If I understand your comment...
Maybe we should define dedicated deltaIPhi and deltaIPhi for the outer loop and make these configurable.
Do I understand correctly your point? thanks

Reply:
@amartelli I understand the logic, but from preliminary time-measurement I did on TTbar+200PU, the time taken by this iteration is a little too much.
I'd suggest for the moment to restrict the region to the single bin and pay, in case, some penalty on the efficiency side. Unless you want to give a try with a fixed 3x3 tower.

Reply:
changed to +/-1 by default for all iterations at the moment.